File: AgoraChatSDK/CWrapper/Wrapper/Wrapper_Android/JavaWrapperDemo/wrapper/src/main/java/com/hyphenate/wrapper/EMMessageManager.java
package com.hyphenate.wrapper;

import com.hyphenate.chat.EMClient;
import com.hyphenate.chat.EMMessage;
import com.hyphenate.chat.EMMessagePinInfo;
import com.hyphenate.chat.EMMessageReaction;
import com.hyphenate.wrapper.callback.EMWrapperCallback;
import com.hyphenate.wrapper.helper.EMChatThreadHelper;
import com.hyphenate.wrapper.helper.EMMessageReactionHelper;
import com.hyphenate.wrapper.helper.EMPinnedInfoHelper;
import com.hyphenate.wrapper.util.EMHelper;
import com.hyphenate.wrapper.util.EMSDKMethod;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.List;


public class EMMessageManager extends EMBaseWrapper {
    public String onMethodCall(String method, JSONObject jsonObject, EMWrapperCallback callback) throws JSONException {
        String ret = null;
        if (EMSDKMethod.getReactionList.equals(method)) {
            ret = reactionList(jsonObject, callback);
        }else if (EMSDKMethod.groupAckCount.equals(method)){
            ret = getAckCount(jsonObject, callback);
        }else if (EMSDKMethod.getChatThread.equals(method)) {
            ret = getChatThread(jsonObject, callback);
        }else if (EMSDKMethod.pinnedInfo.equals(method)) {
            ret = getPinnedInfo(jsonObject, callback);
        }else {
            ret = super.onMethodCall(method, jsonObject, callback);
        }
        return ret;
    }

    private String reactionList(JSONObject params, EMWrapperCallback callback) throws JSONException {
        String msgId = params.getString("msgId");
        EMMessage msg = getMessageWithId(msgId);
        JSONArray jsonArray = new JSONArray();
        if (msg != null) {
            List<EMMessageReaction> reactions = msg.getMessageReaction();
            for (int i = 0; i < reactions.size(); i++) {
                jsonArray.put(EMMessageReactionHelper.toJson(reactions.get(i)));
            }
        }
        return EMHelper.getReturnJsonObject(jsonArray).toString();
    }


    private String getAckCount(JSONObject params, EMWrapperCallback callback) throws JSONException {
        String msgId = params.getString("msgId");
        EMMessage msg = getMessageWithId(msgId);
        return EMHelper.getReturnJsonObject(msg != null ? msg.groupAckCount() : 0).toString();
    }

    private String getChatThread(JSONObject params, EMWrapperCallback callback) throws JSONException {
        String msgId = params.getString("msgId");
        EMMessage msg = getMessageWithId(msgId);
        return EMHelper.getReturnJsonObject(EMChatThreadHelper.toJson(msg.getChatThread())).toString();
    }

    private EMMessage getMessageWithId(String msgId) {
        return EMClient.getInstance().chatManager().getMessage(msgId);
    }

    private String getPinnedInfo(JSONObject params, EMWrapperCallback callback) throws JSONException {
        String msgId = params.getString("msgId");
        final EMMessage msg = EMClient.getInstance().chatManager().getMessage(msgId);
        String ret = "";
        if (null != msg) {
            EMMessagePinInfo pinnedInfo = msg.pinnedInfo();
            if (null != pinnedInfo) {
                ret = EMHelper.getReturnJsonObject(EMPinnedInfoHelper.toJson(pinnedInfo)).toString();
            }
        }
        return ret;
    }
}

File: AgoraChatSDK/CWrapper/Wrapper/Wrapper_iOS/wrapper/wrapper/managers/EMMessageManager.m
//
//  EMMessageManager.m
//  wrapper
//
//  Created by 杜洁鹏 on 2022/11/15.
//

#import "EMMessageManager.h"
#import "EMMessageReaction+Helper.h"
#import "EMChatThread+Helper.h"
#import "EMMessagePinInfo+Helper.h"
#import "EMHelper.h"
#import "EMUtil.h"

@implementation EMMessageManager

- (instancetype)init {
    if (self = [super init]) {
        [self registerEaseListener];
    }
    
    return self;
}

- (NSString *)onMethodCall:(NSString *)method
                    params:(NSDictionary *)params
                  callback:(EMWrapperCallback *)callback {
    
    NSString *ret = nil;
    if ([getReactionList isEqualToString:method]) {
        ret = [self getReactionList:params callback:callback];
    } else if ([groupAckCount isEqualToString:method]) {
        ret = [self groupAckCount:params callback:callback];
    } else if ([getChatThread isEqualToString:method]) {
        ret = [self getChatThread:params callback:callback];
    } else if ([getPinnedInfo isEqualToString:method]) {
        ret = [self getPinnedInfo:params callback:callback];
    } else {
        ret = [super onMethodCall:method params:params callback:callback];
    }
    return ret;
}

- (NSString *)getReactionList:(NSDictionary *)params callback:(EMWrapperCallback *)callback {
    EMChatMessage *msg = [self getMessage:params];
    NSMutableArray *list = [NSMutableArray array];
    for (EMMessageReaction *reaction in msg.reactionList) {
        [list addObject:[reaction toJson]];
    }
    
    return [[EMHelper getReturnJsonObject:list] toJsonString];
}

- (NSString *)groupAckCount:(NSDictionary *)params callback:(EMWrapperCallback *)callback {
    EMChatMessage *msg = [self getMessage:params];
    return [[EMHelper getReturnJsonObject:@(msg.groupAckCount)] toJsonString];
}

- (NSString *)getChatThread:(NSDictionary *)params callback:(EMWrapperCallback *)callback {
    EMChatMessage *msg = [self getMessage:params];
    return [[EMHelper getReturnJsonObject:[msg.chatThread toJson]] toJsonString];
}

- (EMChatMessage *)getMessage:(NSDictionary *)params {
    NSString *msgId = params[@"msgId"];
    EMChatMessage *msg = [EMClient.sharedClient.chatManager getMessageWithMessageId:msgId];
    return msg;
}

- (NSString *)getPinnedInfo:(NSDictionary *)param
                           callback:(EMWrapperCallback *)callback {
    NSString *msgId = param[@"msgId"];

    EMChatMessage *msg = [EMClient.sharedClient.chatManager getMessageWithMessageId:msgId];
    if (msg) {
        if (msg.pinnedInfo) {
            return [[EMHelper getReturnJsonObject:[msg.pinnedInfo toJson]] toJsonString];
        }
    }

    return nil;
}

- (void)registerEaseListener{
    
}
@end
